### DISCUSSION TITLE
Announcing the Beta Release of nextjs-auth0 SDK v4 🎉

### DISCUSSION THREAD

**Discussion by brth31** [auth0/nextjs-auth0#2110] [General]
Hello everyone,

We're thrilled to announce the beta release of nextjs-auth0 SDK v4! This new version brings significant improvements, new features, and fixes to enhance your development experience.

### ⚠️ Important Notice About v3
As we move forward, **we will not be updating v3 of the SDK to support Next.js 15**. This allows us to focus on v4, which offers a wealth of new features and improvements. This will also enable us to support future releases of Next.js faster and with more confidence. We understand this may pose challenges, and we're here to help.

v3 will continue to receive critical security updates for 6 months after the GA of v4.

### 📣   Highlights of v4 Beta
- Middleware-Based Authentication: Improved compatibility and reduced maintenance by moving to middleware-based handlers.
- Enhanced Security: Switched to encrypted cookies and removed outdated cookie logic.
- Resolved State Mismatch Issues: Fixed long-standing issues reported by the community.
- Improved Session Management: Implemented rolling sessions and eliminated cookie chunking.
- Improved Hooks and Helpers: Introduced useUser(), getAccessToken(), and getSession() for easier data fetching and session handling.
- Stateful Sessions with Custom Databases: Support for "Bring Your Own Database" (BYODB).
- Compatibility with Next.js 15, Turbopack, and React 19
- Simplified architecture, API, and configuration options

### ✍️  Try It Out and Provide Feedback
We invite you to explore the beta release and share your feedback to help us improve before the general availability release. We are currently targeting a general availability release by the end of December. 

**Beta Release:**[ v4.0.0-beta.3](https://github.com/auth0/nextjs-auth0/releases/tag/v4.0.0-beta.3)

### ⛵ Need Help with Migration?
If you encounter challenges migrating to v4, please don't hesitate to[ open an issue](https://github.com/auth0/nextjs-auth0/issues) and our team will assist you. We're committed to making the transition as smooth as possible.

Thank You for Your Support 🙌 

We appreciate your understanding as we focus on making v4 the best it can be. Your feedback is invaluable, and we're here to support you every step of the way.

Happy coding! 🚀

— The Auth0 DX SDK Team

**Comment by curry1337** [comment]
Thank you for the beta release! Do you have any migration documentation or best practices available?

**Comment by brth31** [comment]
@curry1337 we do have a comprehensive documentation that you can refer to- https://github.com/auth0/nextjs-auth0/blob/v4/README.md. Hope this helps!

**Reply by jeltehomminga** [reply]
file not found :/

**Comment by johncarmack1984** [comment]
@brth31 The documentation is lacking a coherent migration strategy for those using earlier versions; so far advice from the only auth0 engineer in these threads has delivered [application-breaking fixes that actually made our application less secure](https://github.com/auth0/nextjs-auth0/issues/1790#issuecomment-2468952083). I would disagree with the use of the word "comprehensive" in your previous comment, and I disagree with a product strategy that ignores [what paying customers are telling you](https://github.com/auth0/nextjs-auth0/issues/1779#issuecomment-2481971431) about this release's clear unreadiness for use in production environments.

**Comment by curry1337** [comment]
@brth31 Thanks, what about the Edge Runtime? Couldn't find anything about it in the Readme.
We now use @auth0/nextjs-auth0/edge in the middleware.

Is _import { Auth0Client } from "@auth0/nextjs-auth0/server"_ replacing the old edge?

**Comment by johncarmack1984** [comment]
@curry1337 not representing auth0 or Vercel, but happy to help a fellow engineer: [Vercel reverted all edge rendering back to Node.js seven months ago](https://x.com/leeerob/status/1780705942734331983?lang=en), so the distinction is currently moot from a practical standpoint (although the [Edge API remains in the Next.js docs](https://nextjs.org/docs/app/building-your-application/rendering/edge-and-nodejs-runtimes), so Edge rendering may return in time, so it makes sense you would want the auth0 adapter to ensure support for it). Hope this helps you and your team plan for the future!

**Comment by curry1337** [comment]
@johncarmack1984 Wow thanks so much for sharing, really appreciated!

**Comment by mitchheddles** [comment]
Is there, or will there be, a way to use this without middleware? Middleware only feels like more of a limitation, rather than a "feature".

**Update:** After some testing, it seems like we're able to use the `export const GET = auth0Client.middleware;` in a route handler, as long as the routes match, e.g. `/src/app/auth/[auth0]/route.ts`. It doesn't feel right, but it's good enough for now.

**Comment by portal7** [comment]
I have tried (albeit a bit “commending myself to god”) to implement from 0 in a totally clean project using the initial steps indicated in the “getting started” article of NextJS 15.

I followed the steps they mention in v4 beta, but not only is it not entirely clear, but it didn't work for me.
as you can see in the attached image.

Is there any place where you can give a “practical” example or a test repo, or a sandbox to see how you implement it?
Thanks

![image](https://github.com/user-attachments/assets/75690cbc-d727-4620-8b38-2a3b8aff718d)

**Comment by MrUprizing** [comment]
Hi @portal7, I have a project that will be deployed to NextJs 15, and the Auth0 SDK is delaying us. We're going to wait for version 1.0 of SDKv4. In the meantime, I created a basic example to see how it works. You can check it out to get an idea.
https://github.com/uprizingFaze/NextJs_15-Auth0_V4

**Comment by portal7** [comment]
> Hi @portal7, I have a project that will be deployed to NextJs 15, and the Auth0 SDK is delaying us. We're going to wait for version 1.0 of SDKv4. In the meantime, I created a basic example to see how it works. You can check it out to get an idea. https://github.com/uprizingFaze/NextJs_15-Auth0_V4

Thanks!!! I really appreciate it, I'm going to try it out and then.... we'll see.

**Comment by portal7** [comment]
Well, I downloaded it and tried to run it, but I have not been able to get it to work.

(unfortunately, I don't have enough experience in JS development to be able to go deeper and investigate what would be the problem).
![image](https://github.com/user-attachments/assets/f5076968-15e2-432a-830e-17f23e85eb03)

**Comment by MrUprizing** [comment]
Oh, this is because of npm i, many packages need to specify the version of React 19 to avoid that issue. 
Just use `npm i --force` when using Next.js 15 or any React 19 project (if you use pnpm, Bun, or Yarn, I think this isn't necessary). [Here’s a brief explanation](https://ui.shadcn.com/docs/react-19)

**Comment by portal7** [comment]
@uprizingFaze , thanks for the clarification, I will read the explanation to improve my knoweldge.

After rename and fil the values inside the .env.local values was able to run it with any other console error.

But....
![image](https://github.com/user-attachments/assets/0138ca5d-a1d1-4e11-a4a5-c3a9be740075)


Same as my capture with a fresh-started project with the new package installed.

**Comment by MrUprizing** [comment]
It is likely due to the Auth0 environment variables or the URLs in the Auth0 dashboard.

[Check this](https://github.com/auth0/nextjs-auth0/tree/v4?tab=readme-ov-file#2-add-the-environment-variables)

Dashboard of my example:
<img width="1013" alt="Captura de pantalla 2024-11-19 a la(s) 7 34 10 p m" src="https://github.com/user-attachments/assets/aa67be12-dba6-4758-802b-db9335de0eb0">

**Comment by portal7** [comment]
Well, after configuring the variables again in auth0 with the indicated values (I registered a new app so as not to modify what I had) and **it worked.**

**The strange thing was that** when I clicked on logout, I got an error that the value of the logout URL that I had passed was not valid.
![image](https://github.com/user-attachments/assets/b50417cb-9200-4e01-b8d4-09ca6d1fa155)

And this is my configuration on Auth0
![image](https://github.com/user-attachments/assets/42512817-b2ea-4920-89b7-2c917f1924a4)

**this was fixed when I added** http://localhost:3000 (what the Auth0 error page said) as an enabled value.
Note: I am using beta-3 version, I don't know if this is solved in version 7, but I will try it.

![image](https://github.com/user-attachments/assets/c4899e1c-27fb-4ba6-a2b0-dbae905c0887)


Thanks @uprizingFaze 
I strongly appreciate your time and the help/guidance you have given me.

**Comment by ctcooper** [comment]
@portal7 Did you make sure your app in Auth0 is set as a 'Regular Application'?

'Settings -> Application Properties -> Application Type -> Regular Web Application'
![Screenshot 2024-11-20 at 9 06 15 AM](https://github.com/user-attachments/assets/cfba28d9-6670-4c89-b6d5-b144ea20ee72)

**Comment by portal7** [comment]
Hi @ctcooper , 
No, the value selected in Auth0 was Single Page (SPA)

**Comment by paradimes** [comment]
> Hi @ctcooper , No, the value selected in Auth0 was Single Page (SPA)

for next, you should use Regular Web Application:
<img width="1029" alt="Screenshot 2024-11-20 at 3 59 15 PM" src="https://github.com/user-attachments/assets/2a467f2b-2ed8-4926-9464-3fba31c3f88b">

**Comment by jamiebytebender** [comment]
Is there any example of the new v4 working with org invitations? 
I have a solution similar to this https://github.com/auth0/nextjs-auth0/issues/701#issuecomment-2293009478 where I was calling `handleLogin` previously  but on the latest version there is no `handleLogin` exported.

**Comment by willedanielsson** [comment]
> @curry1337 we do have a comprehensive documentation that you can refer to- https://github.com/auth0/nextjs-auth0/blob/v4/README.md. Hope this helps!

Hey, any news on a migration guide? 
Like 
"Rename env variable XX to YY" etcetera so it's easy to migrate and that people can propose updates to? 
As someone going from 3 to 4 due to requirement for Nextjs, it would be good with a migration guide and not an installation guide 🙌

**Comment by curry1337** [comment]
@brth31  Could you please provide an update on when we can expect the **stable release** of v4? Our team has decided to wait for the stable release and not rely on the beta version.

**Comment by brth31** [comment]
@curry1337 we are tentatively aiming to go GA later this month. I will share more updates closer to the release date.

**Comment by curry1337** [comment]
@brth31 Great! Thank you for replying so quickly.

**Comment by neptaco** [comment]
When I retrieve the AccessToken, the Payload section is missing. Is this issue occurring for anyone else? This happens both when setting up from scratch with beta9 and accessing /auth/access-token, as well as in projects updated from v3.5.0. In both cases, the Payload section is missing.

**Comment by mcgaryes** [comment]
@neptaco - I was not seeing the payload until I overwrote the `beforeSessionSaved` property. I also changed my `audience` in `authorizationParameters` to get the accessToken back in a JWT format. 

```
export const auth0 = new Auth0Client({
    async beforeSessionSaved(session) {
        return {
            ...session,
            user: {
                ...session.user,
            },
        }
    },
    authorizationParameters: {
        audience: "https://service.com", // service that I was needing a JWT for 
    },
});
```

**Comment by neptaco** [comment]
@mcgaryes 
Thank you for the information!
I tried it, and in my case, I was able to retrieve the information just by setting the audience.

In v3.5.0, I used the environment variable AUTH0_AUDIENCE for the configuration, but in v4, it seems that this setting is no longer applied via environment variables.

I’m not sure whether this change is intentional in v4 or if it’s a bug where environment variables are not being loaded.

**Comment by chris-erickson** [comment]
> @mcgaryes Thank you for the information! I tried it, and in my case, I was able to retrieve the information just by setting the audience.
> 
> In v3.5.0, I used the environment variable AUTH0_AUDIENCE for the configuration, but in v4, it seems that this setting is no longer applied via environment variables.
> 
> I’m not sure whether this change is intentional in v4 or if it’s a bug where environment variables are not being loaded.

Docs seem to indicate this is now configured with manually setting up the client. Would be good for the migration one to indicate this needs to be adjusted if so!

~~Wondering as well if it's contributing to this error (that we're also seeing, but slightly different, can open a new issue as needed): https://github.com/auth0/nextjs-auth0/issues/1841#issuecomment-2540811430~~

`TypeError: Failed to parse URL from /auth/access-token`. 

Update: This seems to be not influencing to aforementioned bug. https://github.com/auth0/nextjs-auth0/issues/1843

**Comment by jdwitten** [comment]
I also stumbled on the audience issue. Needed to supply it in the auth0 client initialization like @mcgaryes mentioned to get a valid payload in the JWT access token. Would be good to document this in the migration guide if it is intended.

**Comment by isoroka-plana** [comment]
We noticed that with v4 Auth0 uses `/oidc/logout` instead of `/v2/logout` which doesn't work for us for some reason and the error is not detailed at all. Logout just fails with error `400` and in Auth0 logs says something like "Failed to logout user".
Is this configurable somehow or the oidc logout is now the main? There is no mention of this change in the migration guide nor v4 readme.

EDIT: `/oidc/logout` works if we don't provide `returnTo` which is not registered as Allowed Logout URL, so that was a different issue, found a workaround for it.

**Comment by jonkwheeler** [comment]
> Is there, or will there be, a way to use this without middleware?

Furthermore, right now you cannot use Middleware with OpenNext and Cloudflare.

**Comment by calvinbrewer** [comment]
~~I haven't looked deeper at this issue yet and will follow up with more details but running into the following error:~~
<img width="930" alt="image" src="https://github.com/user-attachments/assets/5d58c4ee-9258-42e4-a1e3-498a6b43db6d" />

Next.js version: 15.1.0
React version: 19.0.0
@auth0/nextjs-auth0: 4.0.0-beta.11

Update: The Auth0 application must have at least one connection. The actual error was highlighted in the OAuth exchange.

**Comment by jamiebytebender** [comment]
Any updates on how organization invites are used and implemented with v4?

**Comment by brandanmajeske** [comment]
_My kingdom for a migration guide._ I've followed the Beta readme and am not succeeding. So many breaking changes, I think I need to see a bare bones example project from the authors.

**Comment by jonkwheeler** [comment]
I'm almost ready to give the node package a try...

https://github.com/auth0/node-auth0
https://auth0.github.io/node-auth0/index.html

In the time I've spent debugging auth0 for nextjs I might have been able to built my own auth provider. Some people found it easy. Others didn't.

**Comment by isoroka-plana** [comment]
Is there any new GA target date?

**Comment by JordanHoffman** [comment]
Is there something I need to do to get useUser() to stop pinging /auth/profile when I'm logged out. It seems like every 10 seconds it pings it even though it's unauthorized (I don't know if this is a bug or I'm doing something wrong, I'm literally just calling useUser()):

![image](https://github.com/user-attachments/assets/825791e4-18bd-4a5f-94ca-6ec29caae58f)

Edit, I think this has to do with the way I log out. In order for logout to work, I enabled RP-Initiated Logout End Session Endpoint Discovery. I notice that this unwanted pinging only starts happening after I first log out via a logout anchor url.

But I still don't know whats being done wrong or how to fix it.

**Comment by jonschmidt** [comment]
> ~I haven't looked deeper at this issue yet and will follow up with more details but running into the following error:~ <img alt="image" width="930" src="https://private-user-images.githubusercontent.com/10352306/396704751-5d58c4ee-9258-42e4-a1e3-498a6b43db6d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzMzA4MjYsIm5iZiI6MTczNjMzMDUyNiwicGF0aCI6Ii8xMDM1MjMwNi8zOTY3MDQ3NTEtNWQ1OGM0ZWUtOTI1OC00MmU0LWExZTMtNDk4YTZiNDNkYjZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDEwMDIwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFjMmM0YWJkNThmYWJiNTEwY2EzODc5ZjYxOWQxZGNiOTVmY2U3YjhhOWVjODU3OTQwN2ExZDZkMWFiYTExNTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.7sUh1jtc53jFmLhzqWaucimIJ4zsQmvS5A6v7nJIWBA">
> 
> Next.js version: 15.1.0 React version: 19.0.0 @auth0/nextjs-auth0: 4.0.0-beta.11
> 
> Update: The Auth0 application must have at least one connection. The actual error was highlighted in the OAuth exchange.

I am getting this error, but I am using an application that in fact has a connection, so this is not the problem I think. Could it be something else? Is it possible that a cookie from v3 is not compatible with a cookie from v4? I am currently logged in with a user that was logged in using v3 while testing the upgrade. It would obviously be very bad if everyone had to logout to accomplish this upgrade.


Edit: It turns out this was because the cookie shares the same name as the Clerk cookie, which is "__session". Deleting that session cookie fixed the problem. However it would be nice to be able to configure the cookie name!

**Comment by kylemorena** [comment]
Hello, does the latest version support Preview URLs and Branch URLs on Vercel?

The issue is having different domains, causing the redirect URI to mismatch.

**Comment by EvanLovely** [comment]
> Hello, does the latest version support Preview URLs and Branch URLs on Vercel?
> 
> The issue is having different domains, causing the redirect URI to mismatch.

Use a `*` in the domain to add the Vercel URLs to your Allowed ____ URLs. It'd be something like `https://*-my-project.vercel.app`. Probably smart to not do `https://*.vercel.app` as that'd let any Vercel account use your stuff.

**Comment by jamiebytebender** [comment]
This is getting ridiculous. When is the final version going to be available? What are we paying for? Hundreds of companies paying for these integrations and we are stuck with half baked release.

**Comment by calvinbrewer** [comment]
> This is getting ridiculous. When is the final version going to be available? What are we paying for? Hundreds of companies paying for these integrations and we are stuck with half baked release.

@jamiebytebender I ended up evalutating Clerk as an alternative and it's a much better dev experience.

**Comment by calumerskine** [comment]
Would it be possible to export types from the library, such as the `Auth0ClientOptions`? We're wrapping the initialisation and looking to extend the config.

**Comment by mcgaryes** [comment]
Any thoughts on pulling the `SESSION_COOKIE_NAME` in the `AbstractSessionStore`. I am running into an issue right now where I have two applications on the same tenant that I want to run locally. The problem is that that the `azp` keeps getting swapped from one client ID to the other, which breaks the jw decrypt functionality. 

My original thought was to update my etc/hosts file, but this comes with its own challenges. It wont be an issue when we're on separate domains, but would be a huge DX improvement when developing locally.

There are some other places that pull in environment variables so I think it would be a simple as:

```
public SESSION_COOKIE_NAME = process.env.SESSION_COOKIE_NAME ?? "__session"

```

**Comment by mattmchugh4** [comment]
Any update on the timeline for the stable release of v4?

**Comment by jonschmidt** [comment]
> Is there something I need to do to get useUser() to stop pinging /auth/profile when I'm logged out. It seems like every 10 seconds it pings it even though it's unauthorized (I don't know if this is a bug or I'm doing something wrong, I'm literally just calling useUser()):
> 
> ![image](https://github.com/user-attachments/assets/825791e4-18bd-4a5f-94ca-6ec29caae58f)
> 
> Edit, I think this has to do with the way I log out. In order for logout to work, I enabled RP-Initiated Logout End Session Endpoint Discovery. I notice that this unwanted pinging only starts happening after I first log out via a logout anchor url.
> 
> But I still don't know whats being done wrong or how to fix it.

@JordanHoffman I am noticing this too. Did you ever find a workaround for this?

**Comment by JordanHoffman** [comment]
> > Is there something I need to do to get useUser() to stop pinging /auth/profile when I'm logged out. It seems like every 10 seconds it pings it even though it's unauthorized (I don't know if this is a bug or I'm doing something wrong, I'm literally just calling useUser()):
> > ![image](https://github.com/user-attachments/assets/825791e4-18bd-4a5f-94ca-6ec29caae58f)
> > Edit, I think this has to do with the way I log out. In order for logout to work, I enabled RP-Initiated Logout End Session Endpoint Discovery. I notice that this unwanted pinging only starts happening after I first log out via a logout anchor url.
> > But I still don't know whats being done wrong or how to fix it.
> 
> [@JordanHoffman](https://github.com/JordanHoffman) I am noticing this too. Did you ever find a workaround for this?

Unfortunately no. I'm just working on a personal project so it's not urgent for me, but I'm keeping my fingers crossed that in the final release it will go away. Hopefully some solution can be provided here.

**Comment by jonkwheeler** [comment]
I have a feeling someone in the community is going to make their own community-backed library for Auth0 specifically bc their company selected Auth0 and they can't get out of it. The thought has crossed my mind more than thrice.

**Comment by jamiebytebender** [comment]
> I have a feeling someone in the community is going to make their own community-backed library for Auth0 specifically bc their company selected Auth0 and they can't get out of it. The thought has crossed my mind more than thrice.

It would make sense if Auth0 were free, but we're paying hundreds of dollars for it. We specifically chose Auth0 because it's a paid service and with that comes the expectation of better support. This is something Auth0 should be addressing directly - it's frustrating and disappointing that they're so unresponsive.

**Comment by jdwitten** [comment]
Is there an updated expected release date for v4? It's been almost a month since the original announced release date has passed and multiple requests for an update. At least an update on what the status is would be appreciated. All of our other 3rd party sdks and libraries have pretty quickly updated to support React 19/ Next 15, but we are blocked on the upgrade due to this release.

**Comment by aeoh31** [comment]
@brth31 can we get an update on the release date please?

**Comment by brth31** [comment]
@jdwitten apologies for the delay. Security review and other release readiness tasks are taking us more time than we anticipated. We are unable to rush through these tasks considering this is a major version release with significant changes to the SDK internals. 

We're trying out best to release the GA this week. I appreciate your patience here.

**Comment by jdwitten** [comment]
Thanks for the update @brth31 👍

**Comment by chammond3** [comment]
I would love a vanilla example of v4 working properly. I'm following the steps in the v4 fork and struggling to get an error of `Error [TypeError]: Invalid URL` from `return await auth0.middleware(request)` in middleware. 

<img width="1000" alt="Image" src="https://github.com/user-attachments/assets/6947379c-f05c-4a2c-9859-a7164b2f3e21" />

I have a basic nextjs project using the app router that I recently built and trying to prove out auth0. I had v3 working with the exception of the [getSession breaking change](https://community.auth0.com/t/nextjs-15-breaking-change-using-getsession-and-touchsession/150741) in next 15. I'm really spinning my wheels converting to v4...

**Comment by brth31** [comment]
🎉 [v4](https://github.com/auth0/nextjs-auth0/releases/tag/v4.0.0) is now GA. Thanks for your patience, everybody!

**Comment by kylemorena** [comment]
> > Hello, does the latest version support Preview URLs and Branch URLs on Vercel?
> > The issue is having different domains, causing the redirect URI to mismatch.
> 
> Use a `*` in the domain to add the Vercel URLs to your Allowed ____ URLs. It'd be something like `https://*-my-project.vercel.app`. Probably smart to not do `https://*.vercel.app` as that'd let any Vercel account use your stuff.

Hi, 
thanks for replying!
 the issue I'm facing is managing VERCEL_BRANCH_URL and VERCEL_URL. I'm using VERCEL_URL as the appBaseUrl. Is there an alternative way to handle this? In version 3, I was able to build the redirectUri by retrieving the request host([request.nextUrl.origin](https://nextjs.org/docs/app/api-reference/functions/next-request#nexturl)) from the API route handler in Next.js.

**Comment by jamiebytebender** [comment]
I have similar issue @kylemorena with preview deployments. 
I was also using the following in the .env.production `NEXT_PUBLIC_AUTH0_BASE_URL=${VERCEL_URL:-$VERCEL_BRANCH_URL}` but doesn't seem to work anymore.

**Comment by KrustyC** [comment]
> 🎉 [v4](https://github.com/auth0/nextjs-auth0/releases/tag/v4.0.0) is now GA. Thanks for your patience, everybody!
@brth31 Thanks for getting this over the line. I have a question though:

Is there a plan to actually release this on NPM as well? Seems like the [latest available version on NPM](https://www.npmjs.com/package/@auth0/nextjs-auth0/v/4.0.0-beta.14) is not an official one, but it's just a beta?

**Comment by isoroka-plana** [comment]
> > 🎉 [v4](https://github.com/auth0/nextjs-auth0/releases/tag/v4.0.0) is now GA. Thanks for your patience, everybody!
> > [@brth31](https://github.com/brth31) Thanks for getting this over the line. I have a question though:
> 
> Is there a plan to actually release this on NPM as well? Seems like the [latest available version on NPM](https://www.npmjs.com/package/@auth0/nextjs-auth0/v/4.0.0-beta.14) is not an official one, but it's just a beta?

@KrustyC there's 4.0.0 in the version history so I am pretty sure you can install it.

![Image](https://github.com/user-attachments/assets/1895a3a2-6a54-4c1c-a751-b8b711e4278a)
Question is why is it not among the current tags instead of the beta one?

**Comment by chris-erickson** [comment]
It's also still only in the v4 branch instead of main which is incredibly confusing when coming back here to look for v4 docs. Takes some time digging into PR's to really even be sure _that's_ that correct branch.

**Comment by curry1337** [comment]
Fellow developers who have migrated: What’s your advice for others? Should we move to v4 now or hold off? We’re planning to upgrade to Next.js 15 but want to avoid debugging Auth0, as our roadmap is already packed.

**Comment by chris-erickson** [comment]
Would reccomend. Overall it works much more like you'd expect and no major issues detected here. Glad to have NextJS 15 features.

**Comment by curry1337** [comment]
> Would reccomend. Overall it works much more like you'd expect and no major issues detected here. Glad to have NextJS 15 features.

Thank you!

**Comment by DotCoyote** [comment]
> Fellow developers who have migrated: What’s your advice for others? Should we move to v4 now or hold off? We’re planning to upgrade to Next.js 15 but want to avoid debugging Auth0, as our roadmap is already packed.

Overall it works quite good, I just ran into an edgecase: https://github.com/auth0/nextjs-auth0/issues/1884

**Comment by perryraskin** [comment]
@chammond3 I'm having the same issue as you - Invalid URL happening in the middleware file. I'm on v4.0.1. Is there a solution? I have the correct env variables so I'm not sure what the issue is.

**UPDATE:** It appears that changing the env var from `AUTH0_BASE_URL` to `APP_BASE_URL` did the trick.

**Comment by mmeadwell** [comment]
Has anyone implemented a custom session store using redis or a database? Reading through the documentation I see how to setup the custom session store but not any implementation details. Thanks for any help in advance!

**Comment by jonkwheeler** [comment]
Excited about the release. Anyone else having issues with the search bar on the docs site? https://auth0.github.io/nextjs-auth0/modules.html

**Comment by avaldez-gr** [comment]
The amount of changes is substantial with no migration guide. This is extremely frustrating considering my company is spending a substantial amount of money for this. No way to get roles anymore - at least not clearly - because there is no idToken, just an access token. And, even with this, there's no documentation on how to obtain the roles from this.

Cool new features, but the roll out is pretty rough.

**Comment by jonkwheeler** [comment]
Where are the docs for v3?

**Comment by calvinbrewer** [comment]
> Where are the docs for v3?

@jonkwheeler Older README versions can be found on the tagged releases. E.g. https://github.com/auth0/nextjs-auth0/tree/v3.6.0

**Comment by jonkwheeler** [comment]
> > Where are the docs for v3?
> 
> [@jonkwheeler](https://github.com/jonkwheeler) Older README versions can be found on the tagged releases. E.g. https://github.com/auth0/nextjs-auth0/tree/v3.6.0

Yeah, but that's not the full docs. Try following those links. All broke now. 

I'm pretty sure I've exposed that `screen_hint` does nothing. I was trying to find docs on it. I'm stuck just reading the source code instead.

**Comment by jonkwheeler** [comment]
> Cool new features, but the roll out is pretty rough.

Has anyone had a positive experience? Wondering if I attempt the jump now that all v3 support is dropped, and I also can't use Next.js v15 unless I upgrade this now. (OR I fork v3)

**Comment by avaldez-gr** [comment]
> > Cool new features, but the roll out is pretty rough.
> 
> Has anyone had a positive experience? Wondering if I attempt the jump now that all v3 support is dropped, and I also can't use Next.js v15 unless I upgrade this now. (OR I fork v3)

@jonkwheeler we just updated to Next15 and we're forced to use v4. It was painful but we figured it out. It's pretty frustrating there's little documentation on this and the documentation is missing some key items. With v3 you use the api but with v4 you have to use middleware. I do like it better, but that's just because we figured out how to use it. Lots have changed. Can't use the same exported members as before. And have to create your own lib instantiating auth0. 

Metadata is no longer along with the user data. So you have to use the management sdk to get any data from metadata.

**Comment by brentshulman-silkline** [comment]
This is ridiculous. That lack of follow-up and response from the Auth0 team here is borderline negligent. **I'm pissed off, and you should be too.** 

Luckily, the original poster, @brth31, generously gave us his [Calendly link](https://calendly.com/bharath-auth0/product-feedback) on his [Github profile](https://github.com/brth31). I set a meeting with him, and if you feel the same way I do, I encourage you to do the same. 

For example, is this [the migration guide](https://github.com/auth0/nextjs-auth0/blob/v4.0.2/V4_MIGRATION_GUIDE.md) we are looking for? Hell if I know.

**Comment by chris-erickson** [comment]
> > Cool new features, but the roll out is pretty rough.
> 
> Has anyone had a positive experience? Wondering if I attempt the jump now that all v3 support is dropped, and I also can't use Next.js v15 unless I upgrade this now. (OR I fork v3)

Took about a day to migrate a not very complicated site and it's been fine. The communication and documentation has left a _lot_ to be desired and pushed us to seriously explore alternates because it's been so rocky. Just an incredible unforced error.

**Comment by brth31** [comment]
Sincere apologies for the radio silence here 🙏  The team probably had no visibility on the recent comments since the issue was marked as closed. I'm re-opening this issue and paging @tusharpandey13 and @arpit-jn for visibility.

My [calendly link](https://calendly.com/bharath-auth0/product-feedback) is still open for anybody who is willing to share feedback on any of our SDKs. I'm also happy to do working sessions with our engineering team to solve your integration problems. Please add a note while booking the meeting, so that we can come prepared. 

Again, I really appreciate your patience here. We're working really hard to action all the feedback across other GitHub issues on this repo. In the spirit of transparency, here's the prioritised list of problems we're tackling for the upcoming releases: 
1. Server Component Token Management ([#1718](https://github.com/auth0/nextjs-auth0/issues/1718))
2. Native Re-authentication Support ([#1920](https://github.com/auth0/nextjs-auth0/issues/1920) and [#1937](https://github.com/auth0/nextjs-auth0/issues/1937))
3. Improved cookie management with support for built-in cookie splitting mechanisms
4. Support for Partial Prerendering (PPR) 
5. Edge Runtime Support ([#1908](https://github.com/auth0/nextjs-auth0/issues/1908))
6. Improved User Creation Workflow ([#1759](https://github.com/auth0/nextjs-auth0/issues/1759) )


In the upcoming weeks, you should see us making progress on the above issues. Happy to prioritise any other issues that are top of mind for you.

**Comment by brad-decker** [comment]
I have an onboarding modal that only shows up if the useUser hook returns a user with an email property that isn't 'typeof user?.email === 'undefined'' if there is an email, it will popup the modal IF the returned user object from the database doesn't have provided data that is needed by our application. 

What i'm experiencing since adopting 4.1 is that after sometime of being on the site sometimes the modal will pop up for a user - and after refreshing the user is no longer logged in. Its difficult to reproduce cause it requires some kind of session expiration to happen but i'm unsure of what is causing that expiration. Anyone have advice or suggestions on what may be the cause? 

EDIT: Setting 'absolute' and 'inactivity' durations low (1 second) doesn't trigger the same experience locally

**Comment by jonkwheeler** [comment]
In case it helps anyone, we decided we couldn't wait around for Auth0 to fix itself. After looking at WorkOS and Descope, we went to [Descope](https://www.descope.com/). The experience has been quite nice. They have a Slack Community called [AuthTown](https://www.descope.com/community) you can join and ask questions.

**Comment by kzunino** [comment]
Is v4 compatible with Next14? I don't see Next15 as version requirement.

**Reply by arpit-jn** [reply]
Yes, v4 is compatible with Next14 starting with 14.2.25 and Next15 starting with 15.2.3 version.

**Reply by darioscattolini** [reply]
@arpit-jn Is this still so? Want to try the migration before upgrading Next.js to v15, but the [quickstart page](https://auth0.com/docs/quickstart/webapp/nextjs/01-login) says: 

> System requirements: Next.js 15.2.4+
